File: probs/2261.cpp
#include <bits/stdc++.h>

using namespace std;
using pi = pair<int, int>;

const int MAX = 1e5,
          INF = 1e9;

pi A[MAX];

int dist(pi a, pi b)
{
    return (a.first-b.first)*(a.first-b.first) + (a.second-b.second)*(a.second-b.second);
}

int sq(int a)
{
    return a*a;
}

int solve(int l, int h)
{
    if (l >= h) return INF;

    int m = (l+h)/2,
        d = min(solve(l, m), solve(m+1, h));

    vector<pi> tmp; tmp.push_back(A[m]);
    for (int i = m-1; i >= l; i--) {
        if (sq(A[i].first - A[m].first) >= d) break;
        tmp.push_back(A[i]);
    }
    for (int i = m+1; i < h; i++) {
        if (sq(A[i].first - A[m].first) >= d) break;
        tmp.push_back(A[i]);
    }
    sort(tmp.begin(), tmp.end(), [](pi a, pi b) { return a.second == b.second ? a.first < b.first : a.second < b.second; });

    for (int i = 0; i < tmp.size(); i++) {
        for (int j = i+1; j < tmp.size(); j++) {
            if (sq(tmp[j].second - tmp[i].second) >= d) break;
            d = min(d, dist(tmp[i], tmp[j]));
        }
    }

    return d;
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    freopen("input.txt", "r", stdin);

    int N; cin >> N;
    for (int i = 0; i < N; i++) cin >> A[i].first >> A[i].second;
    sort(A, A+N);

    cout << solve(0, N) << '\n';

    return 0;
}

File: probs/2261.py
import sys
input = sys.stdin.readline

input = open('input.txt', 'r').readline

INF = int(1e9)

def sq(n):
    return n*n

def dist(a, b):
    return sq(a[0]-b[0]) + sq(a[1]-b[1])

def solve(l, h):
    if h <= l:
        return INF

    m = (l+h)//2
    d = min(solve(l, m), solve(m+1, h))

    tmp = [A[m]]
    for i in reversed(range(l, m)):
        if sq(A[i][0]-A[m][0]) >= d:
            break
        tmp.append(A[i])
    for i in range(m+1, h):
        if sq(A[i][0]-A[m][0]) >= d:
            break
        tmp.append(A[i])
    tmp.sort(key=lambda x: x[::-1])

    for i in range(len(tmp)):
        for j in range(i+1, len(tmp)):
            if sq(tmp[i][1]-tmp[j][1]) >= d:
                break
            d = min(d, dist(tmp[i], tmp[j]))

    return d

N = int(input())
A = [tuple(map(int, input().split())) for _ in range(N)]
A.sort()

print(solve(0, N))
